src/protocol/src/sql_values.rs

```diff
@@ -0,0 +1,326 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::{sql_formats::PostgreSqlFormat, sql_types::PostgreSqlType};
+use byteorder::{BigEndian, ReadBytesExt};
+use std::str;
+
+type Result = std::result::Result<PostgreSqlValue, String>;
+
+/// Represents PostgreSQL data values sent and received over wire
+#[allow(missing_docs)]
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum PostgreSqlValue {
+    Null,
+    True,
+    False,
+    Int16(i16),
+    Int32(i32),
+    Int64(i64),
+    String(String),
+}
+
+impl PostgreSqlValue {
+    /// Deserializes a value of type `typ` from `raw` using the specified
+    /// `format`.
+    pub fn decode(format: &PostgreSqlFormat, typ: &PostgreSqlType, raw: &[u8]) -> Result {
+        match format {
+            PostgreSqlFormat::Binary => decode_binary(typ, raw),
+            PostgreSqlFormat::Text => decode_text(typ, raw),
+        }
+    }
+}
+
+fn decode_binary(typ: &PostgreSqlType, raw: &[u8]) -> Result {
+    match typ {
+        PostgreSqlType::Bool => parse_bool_from_binary(raw),
+        PostgreSqlType::Char => parse_char_from_binary(raw),
+        PostgreSqlType::VarChar => parse_varchar_from_binary(raw),
+        PostgreSqlType::SmallInt => parse_smallint_from_binary(raw),
+        PostgreSqlType::Integer => parse_integer_from_binary(raw),
+        PostgreSqlType::BigInt => parse_bigint_from_binary(raw),
+        other => Err(format!("Unsupported Postgres type: {:?}", other)),
+    }
+}
+
+fn decode_text(typ: &PostgreSqlType, raw: &[u8]) -> Result {
+    let s = match str::from_utf8(raw) {
+        Ok(s) => s,
+        Err(_) => return Err(format!("Failed to parse UTF8 from: {:?}", raw)),
+    };
+
+    match typ {
+        PostgreSqlType::Bool => parse_bool_from_text(s),
+        PostgreSqlType::Char => parse_char_from_text(s),
+        PostgreSqlType::VarChar => parse_varchar_from_text(s),
+        PostgreSqlType::SmallInt => parse_smallint_from_text(s),
+        PostgreSqlType::Integer => parse_integer_from_text(s),
+        PostgreSqlType::BigInt => parse_bigint_from_text(s),
+        other => Err(format!("Unsupported Postgres type: {:?}", other)),
+    }
+}
+
+fn parse_bigint_from_binary(mut buf: &[u8]) -> Result {
+    let v = match buf.read_i64::<BigEndian>() {
+        Ok(v) => v,
+        Err(_) => return Err(format!("Failed to parse BigInt from: {:?}", buf)),
+    };
+
+    if !buf.is_empty() {
+        return Err("invalid buffer size".into());
+    }
+
+    Ok(PostgreSqlValue::Int64(v))
+}
+
+fn parse_bigint_from_text(s: &str) -> Result {
+    let v: i64 = match s.trim().parse() {
+        Ok(v) => v,
+        Err(_) => return Err(format!("Failed to parse SmallInt from: {}", s)),
+    };
+
+    Ok(PostgreSqlValue::Int64(v))
+}
+
+fn parse_bool_from_binary(buf: &[u8]) -> Result {
+    let len = buf.len();
+    if len != 1 {
```

Comment:
Sorry, I can't see why condition is strict `!= 1`? Can you clarify a bit?

Reply:
I think passing type `bool` and `char` in binary should be validated the length of `[u8]` to one. I referenced the code of `rust-postgres` (which is used by `materialize`): https://github.com/sfackler/rust-postgres/blob/master/postgres-protocol/src/types/mod.rs#L34

Reply:
which `char` is in context? rust single item of a string or `char` sql type that have length?

Reply:
I suppose `char` and `varcher` should be parsed in the same way and `len != 1` is ok for `bool`

Reply:
Is the type `Char` is handled as a single character?

https://github.com/sfackler/rust-postgres/blob/master/postgres-protocol/src/types/mod.rs#L75

Reply:
Sorry. I mean just the SQL type `Char`. It is handled as a single byte when passing in binary (not as `Varchar`).